File: Logitude/LogitudeFrontEnd/LogitudeFrontEnd/app/shipment/EntityPMs/ShipmentPackagePM.ts

import {ShipmentPM} from './ShipmentPM'
import {UIProperties, UIProperty} from '../../infrastructure/logitude-components/UIProperties';

export class ShipmentPackagePM {

    public UIProperties: UIProperties;

    constructor(public entityParentPM: ShipmentPM) {
        this.UIProperties = new UIProperties; 
        this.IsDirty = false;
       
    }

    public IsDirty: boolean;

    private _Id: string;
    public get Id() { return this._Id; }
    public set Id(newValue: string) { this._Id = newValue; this.MarkAsDirty(); }

    
    private _Tenant: number;
    public get Tenant() { return this._Tenant; }
    public set Tenant(newValue: number) { this._Tenant = newValue; this.MarkAsDirty();  }

    private _IsContainer: boolean;
    public get IsContainer() { return this._IsContainer; }
    public set IsContainer(newValue: boolean) { this._IsContainer = newValue; this.MarkAsDirty();  }

    private _PackageTypeId: string;
    public get PackageTypeId() { return this._PackageTypeId; }
    public set PackageTypeId(newValue: string) { this._PackageTypeId = newValue; this.MarkAsDirty(); }

    private _PackageTypeCode: string;
    public get PackageTypeCode() { return this._PackageTypeCode; }
    public set PackageTypeCode(newValue: string) { this._PackageTypeCode = newValue; this.MarkAsDirty(); }

    private _PackageTypeName: string;
    public get PackageTypeName() { return this._PackageTypeName; }
    public set PackageTypeName(newValue: string) { this._PackageTypeName = newValue; this.MarkAsDirty(); }

    private _PrintAs: string;
    public get PrintAs() { return this._PrintAs; }
    public set PrintAs(newValue: string) { this._PrintAs = newValue; this.MarkAsDirty(); }

    private _ContainerSize: number;
    public get ContainerSize() { return this._ContainerSize; }
    public set ContainerSize(newValue: number) { this._ContainerSize = newValue; this.MarkAsDirty(); }

    private _TEU: number;
    public get TEU() { return this._TEU; }
    public set TEU(newValue: number) { this._TEU = newValue; this.MarkAsDirty(); }
    
    private _ContainerNumber: string;
    public get ContainerNumber() { return this._ContainerNumber; }
    public set ContainerNumber(newValue: string) { this._ContainerNumber = newValue; this.MarkAsDirty(); }

    private _MaterialDescription: string;
    public get MaterialDescription() { return this._MaterialDescription; }
    public set MaterialDescription(newValue: string) { this._MaterialDescription = newValue; this.MarkAsDirty(); }


    private _Seal: string;
    public get Seal() { return this._Seal; }
    public set Seal(newValue: string) { this._Seal = newValue; this.MarkAsDirty(); }


    private _Tare: number;
    public get Tare() { return this._Tare; }
    public set Tare(newValue: number) { this._Tare = newValue; this.MarkAsDirty(); }
    
    private _Quantity: number;
    public get Quantity() { return this._Quantity; }
    public set Quantity(newValue: number) { this._Quantity = newValue; this.MarkAsDirty(); }

    private _Weight: number;
    public get Weight() { return this._Weight; }
    public set Weight(newValue: number) { this._Weight = newValue; this.MarkAsDirty(); }
    
    private _Volume: number;
    public get Volume() { return this._Volume; }
    public set Volume(newValue: number) { this._Volume = newValue; this.MarkAsDirty(); }

    private _VolumetricWeight: number;
    public get VolumetricWeight() { return this._VolumetricWeight; }
    public set VolumetricWeight(newValue: number) { this._VolumetricWeight = newValue; this.MarkAsDirty(); }

    private _Height: number;
    public get Height() { return this._Height; }
    public set Height(newValue: number) { this._Height = newValue; this.MarkAsDirty(); }

    private _Width: number;
    public get Width() { return this._Width; }
    public set Width(newValue: number) { this._Width = newValue; this.MarkAsDirty(); }

    private _Length: number;
    public get Length() { return this._Length; }
    public set Length(newValue: number) { this._Length = newValue; this.MarkAsDirty(); }

    private _UnNumber: string;
    public get UnNumber() { return this._UnNumber; }
    public set UnNumber(newValue: string) { this._UnNumber = newValue; this.MarkAsDirty(); }

    private _ClassNumber: string;
    public get ClassNumber() { return this._ClassNumber; }
    public set ClassNumber(newValue: string) { this._ClassNumber = newValue; this.MarkAsDirty(); }

    private _Temperature: number;
    public get Temperature() { return this._Temperature; }
    public set Temperature(newValue: number) { this._Temperature = newValue; this.MarkAsDirty(); }

    private _Ventilation: number;
    public get Ventilation() { return this._Ventilation; }
    public set Ventilation(newValue: number) { this._Ventilation = newValue; this.MarkAsDirty(); }

    private _Seal2: string;
    public get Seal2() { return this._Seal2; }
    public set Seal2(newValue: string) { this._Seal2 = newValue; this.MarkAsDirty(); }

    private _SOC: number;
    public get SOC() { return this._SOC; }
    public set SOC(newValue: number) { this._SOC = newValue; this.MarkAsDirty(); }

    private _PackagingGroup: string;
    public get PackagingGroup() { return this._PackagingGroup; }
    public set PackagingGroup(newValue: string) { this._PackagingGroup = newValue; this.MarkAsDirty(); }

    private _IMDGCode: string;
    public get IMDGCode() { return this._IMDGCode; }
    public set IMDGCode(newValue: string) { this._IMDGCode = newValue; this.MarkAsDirty(); }

    private _FlashPoint: string;
    public get FlashPoint() { return this._FlashPoint; }
    public set FlashPoint(newValue: string) { this._FlashPoint = newValue; this.MarkAsDirty(); }

    private _Harmonize: string;
    public get Harmonize() { return this._Harmonize; }
    public set Harmonize(newValue: string) { this._Harmonize = newValue; this.MarkAsDirty(); }

    private _MarksAndNumbers: string;
    public get MarksAndNumbers() { return this._MarksAndNumbers; }
    public set MarksAndNumbers(newValue: string) { this._MarksAndNumbers = newValue; this.MarkAsDirty(); }

    private _Description: string;
    public get Description() { return this._Description; }
    public set Description(newValue: string) { this._Description = newValue; this.MarkAsDirty(); }

    private _ShipmentNumber: string;
    public get ShipmentNumber() { return this._ShipmentNumber; }
    public set ShipmentNumber(newValue: string) { this._ShipmentNumber = newValue; this.MarkAsDirty(); }

    private _ShipmentId: string;
    public get ShipmentId() { return this._ShipmentId; }
    public set ShipmentId(newValue: string) { this._ShipmentId = newValue; this.MarkAsDirty(); }

    private _ShipmentPMId: string;
    public get ShipmentPMId() { return this._ShipmentPMId; }
    public set ShipmentPMId(newValue: string) { this._ShipmentPMId = newValue; this.MarkAsDirty(); }

    private _IsDangerous: boolean;
    public get IsDangerous() { return this._IsDangerous; }
    public set IsDangerous(newValue: boolean) { this._IsDangerous = newValue; this.MarkAsDirty(); }

    private _OriginalShipmentPackageId: string;
    public get OriginalShipmentPackageId() { return this._OriginalShipmentPackageId; }
    public set OriginalShipmentPackageId(newValue: string) { this._OriginalShipmentPackageId = newValue; this.MarkAsDirty(); }

    private _CommodityId: string;
    public get CommodityId() { return this._CommodityId; }
    public set CommodityId(newValue: string) { this._CommodityId = newValue; this.MarkAsDirty(); }

    private _NumberOfInsidePackages: number;
    public get NumberOfInsidePackages() { return this._NumberOfInsidePackages; }
    public set NumberOfInsidePackages(newValue: number) { this._NumberOfInsidePackages = newValue; this.MarkAsDirty(); }

    private _NumberOfInsidePackagesDetails: string;
    public get NumberOfInsidePackagesDetails() { return this._NumberOfInsidePackagesDetails; }
    public set NumberOfInsidePackagesDetails(newValue: string) { this._NumberOfInsidePackagesDetails = newValue; this.MarkAsDirty(); }

    //Dummy
    private _IsAWBWizardDefault: boolean;
    public get IsAWBWizardDefault() { return this._IsAWBWizardDefault; }
    public set IsAWBWizardDefault(newValue: boolean) { this._IsAWBWizardDefault = newValue; this.MarkAsDirty(); }

    public InsideShipmentPackages: Array<any>;
    public ShipmentPackageItems: Array<any>;

    MarkAsDirty() {
        this.IsDirty = true;
        this.entityParentPM.IsDirty = true;
    }
    
}

File: Logitude/Logitude.BL/ShipmentsModel/EntityPMs/ShipmentPackagePM.cs
using Simplog.Server.Infrastructure;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ServiceModel.DomainServices.Server;
using Logitude.BL.CommonDataModel.EntityPMs;
using System;
using Logitude.BL.Validators;
using Simplog.Data.ShipmentsModel.EntityPOCOs;

namespace Logitude.BL.ShipmentsModel.EntityPMs
{
        
    [CustomValidation(typeof(Validators.ClassLevelValidator), "ValidateClass")]
    [CustomValidation(typeof(ShipmentPackageValidator), "IsShipmentPackageValid")]
    public class ShipmentPackagePM
    {
        [Key]
        public string Id { get; set; }
        public int Tenant { get; set; }
        public bool IsContainer { get; set; }
        public bool IsContainerRefrigerated { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public string PackageTypeId { get; set; }
        public string PackageTypeCode { get; set; }
        public string PackageTypeName { get; set; }
        public string PrintAs { get; set; }
        public int ContainerSize { get; set; }
        public double TEU { get; set; }

        public bool PackageTypeIsAir { get; set; }
        public bool PackageTypeIsOcean { get; set; }
        public bool PackageTypeIsInland { get; set; }
        public string PackageTypeNote { get; set; }
        public string PackageTypeLocalName { get; set; }
        public decimal PackageTypeVolume { get; set; }
        public bool IsPackageAddedManually { get; set; }
        public string WarehouseReleaseNumber { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public string ContainerNumber { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public string MaterialDescription { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public string ShipperSeal { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public double? Tare { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public int? Quantity { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public double? Weight { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public double? Volume { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public double? VolumetricWeight { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public double? Height { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public double? Width { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public double? Length { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public string UnNumber { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public string ClassNumber { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public string Temperature { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public double? Ventilation { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public string CarrierSeal { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public int? SOC { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public string PackagingGroup { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public string IMDGCode { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public string FlashPoint { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public string Harmonize { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public string MarksAndNumbers { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public string Description { get; set; }

        public string ShipmentNumber { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public string ShipmentId { get; set; }       
        public string ShipmentPMId { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public bool IsDangerous { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public string OriginalShipmentPackageId { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public string CommodityId { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public int NumberOfInsidePackages { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public string NumberOfInsidePackagesDetails { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public decimal? VGM { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public string MethodUsed { get; set; }

        public bool IsDeliveryFU { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public string DeliveryId { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public DateTime? DeliveryETD { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public DateTime? DeliveryATD { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public DateTime? DeliveryETA { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public DateTime? DeliveryATA { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public string DeliveryTo { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public string DeliveryFrom { get; set; }

        public bool IsEmptyContainerReturnFU { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public string EmptyContainerReturnId { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public DateTime? EmptyContainerReturnETD { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public DateTime? EmptyContainerReturnATD { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public DateTime? EmptyContainerReturnETA { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public DateTime? EmptyContainerReturnATA { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public string EmptyContainerReturnTo { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public string EmptyContainerReturnFrom { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public string CeficClass { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public string KelmerCode { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public string EMS { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public string ProperShippingName { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public bool MarinePollutant { get; set; }

        public bool NonActiveContainer { get; set; }

        //Dummy
        public bool IsAWBWizardDefault { get; set; }
        public string DummyIdGuid { get; set; }        

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public DateTime? OnCarriageETD { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public DateTime? OnCarriageATD { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public DateTime? OnCarriageETA { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public DateTime? OnCarriageATA { get; set; }


        // public string CustomerId { get; set; }
        //public string DirectionId { get; set; }
        //public string MainCarriageCarrierNumber { get; set; }
        // public DateTime CreateDateTime { get; set; }
        //public DateTime? MainCarriageATD { get; set; }
        //public string BookingConfirmationNumber { get; set; }
        //public string MainCarriageFromPortCode { get; set; }
        //public string MainCarriageFinalDestinationPortCode{ get; set; }

        [Include]
        [Association("ShipmentPackagePMShipment1", "ShipmentPMId", "Id", IsForeignKey = true)]
        public virtual ShipmentPM ShipmentPM { get; set; }

        private List<InsideShipmentPackagePM> insideShipmentPackages;
        [Composition]
        [Include]
        [Association("ShipmentPackageInsideShipmentPackage", "Id", "ShipmentPackageId")]
        public virtual List<InsideShipmentPackagePM> InsideShipmentPackages
        {
            get
            {
                if (insideShipmentPackages == null)
                {
                    insideShipmentPackages = new List<InsideShipmentPackagePM>();
                }
                return insideShipmentPackages;
            }
            set { insideShipmentPackages = value; }
        }

        private List<ShipmentPackageItemPM> shipmentPackageItems;
        [Composition]
        [Include]
        [Association("ShipmentPackageShipmentPackageItem", "Id", "PackageId")]
        public virtual List<ShipmentPackageItemPM> ShipmentPackageItems
        {
            get
            {
                if (shipmentPackageItems == null)
                {
                    shipmentPackageItems = new List<ShipmentPackageItemPM>();
                }

                return shipmentPackageItems;
            }

            set { shipmentPackageItems = value; }
        }

        public ChangeSetOperation ChangeSetOp { get; set; }

        private List<InsideShipmentPackagePM> insideShipmentPackagesChangeSet;
        public List<InsideShipmentPackagePM> InsideShipmentPackagesChangeSet
        {
            get
            {
                if (insideShipmentPackagesChangeSet == null)
                {
                    insideShipmentPackagesChangeSet = new List<InsideShipmentPackagePM>();
                }

                return insideShipmentPackagesChangeSet;
            }

            set
            {
                insideShipmentPackagesChangeSet = value;
            }
        }

        private List<ShipmentPackageItemPM> shipmentPackageItemsChangeSet;
        public List<ShipmentPackageItemPM> ShipmentPackageItemsChangeSet
        {
            get
            {
                if (shipmentPackageItemsChangeSet == null)
                {
                    shipmentPackageItemsChangeSet = new List<ShipmentPackageItemPM>();
                }

                return shipmentPackageItemsChangeSet;
            }

            set
            {
                shipmentPackageItemsChangeSet = value;
            }
        }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public string Reference1 { get; set; }
        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public string Reference2 { get; set; }
        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public string Reference3 { get; set; }
        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public string Reference4 { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public string CommodityNumber { get; set; }
        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public string CommodityName { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public string Notes { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public string TemperatureUnitCode { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public string FlashPointTemperatureUnitCode { get; set; }

        public int SplitIndex { get; set; }
        public bool IsFromSplit { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public string LastStatusCode { get; set; }

        public string ContainerStatusSourceCode { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public string LastStatusName { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public DateTime? LastStatusDate { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public string DeliveryTransportModeCode { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public string ECRTransportModeCode { get; set; }

        public bool IsMultiHarmonize { get; set; }

        public int InUse { get; set; }

        private List<ShipmentPackageHarmonizePM> shipmentPackageHarmonizes;
        [Composition]
        [Include]
        [Association("ShipmentPackageShipmentPackageHarmonize", "Id", "PackageId")]
        public virtual List<ShipmentPackageHarmonizePM> ShipmentPackageHarmonizes
        {
            get
            {
                if (shipmentPackageHarmonizes == null)
                {
                    shipmentPackageHarmonizes = new List<ShipmentPackageHarmonizePM>();
                }

                return shipmentPackageHarmonizes;
            }

            set { shipmentPackageHarmonizes = value; }
        }

        private List<ShipmentPackageHarmonizePM> shipmentPackageHarmonizesChangeSet;
        public List<ShipmentPackageHarmonizePM> ShipmentPackageHarmonizesChangeSet
        {
            get
            {
                if (shipmentPackageHarmonizesChangeSet == null)
                {
                    shipmentPackageHarmonizesChangeSet = new List<ShipmentPackageHarmonizePM>();
                }

                return shipmentPackageHarmonizesChangeSet;
            }

            set
            {
                shipmentPackageHarmonizesChangeSet = value;
            }
        }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public DateTime? ETD { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public DateTime? ETA { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public string Routing { get; set; }
        public string RoutingIds { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public string VoyageTripNumber { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public bool HasContainerException { get; set; }

        public string Make { get; set; }
        public string Model { get; set; }
        public string Year { get; set; }
        public string Color { get; set; }
        public string ChassisNumber { get; set; }
        public string RegistrationNumber { get; set; }
        public string CountryId { get; set; }
        public string CountryName { get; set; }
        public bool IsVehicle { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public string HorseId { get; set; }
        public string HorseName { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public string LCLContainerTypeId { get; set; }

        [CustomValidation(typeof(Validators.ValidationClass), "ValidateClass")]
        public string ContainerEntityId { get; set; }
    }
}
